### ISSUE TITLE
Create custom @Value.Generated annotation for generated classes with @Retention(CLASS) so JaCoCo can ignore generated classes

### ISSUE THREAD

**Issue by matt-isaman** [immutables/immutables#756]
I opened [an issue](https://github.com/jacoco/jacoco/issues/656) with the JaCoCo folks about ignoring the Immutables generated classes in their code coverage reporting. Their feedback was that they need some way to identify the generated classes in the compiled class files. Apparently Lombok and Groovy both have custom annotations for this purpose (`@lombok.Generated` and `@groovy.transform.Generated`). I think it would be helpful for Immutables to also support this.

**Comment by elucash** [comment]
Thank you for the feature request! So what are the specific requirements, 1. it should be at least class retentions. 2. should it be named 'Generated' (simple name) or some other? Given that we did some work on recognizing `java.annotation.processing.Generated` in 2.6, we can also do something for the custom annotation. Currently, I think it may be a separate module which provides this annotation and auto-activated if placed on the classpath i.e. will use this annotation to mark generated code instead of any other.

**Comment by matt-isaman** [comment]
Yes, requirement 1 is correct, it has to have class retention. I don't think there is a requirement to name the annotation `Generated` but that seems to be the most clear thing to name it and it matches what other code generation libraries have done.

I am not sure if you need to exclude other `Generated` annotations in favor of this new one. I think it would be perfectly fine to add the javax one and this new custom one. So that anything that cared about the javax annotation can continue to pay attention to that one. And anything that cares about the new class retention one can pay attention to it.

So the generated class could look like:
```java
@javax.annotation.Generated({"Immutables.generator", "Foo"})
@org.immutables.value.Value.Generated
public final class ImmutableFoo extends Foo { ... }
```

If you wanted to talk to the JaCoCo guys directly on this you can comment on the issue I opened with them about this: https://github.com/jacoco/jacoco/issues/656

**Comment by gedl** [comment]
👍 to this feature.
Simply changing (or allowing to change via `Style`) the retention of `@Generated` to `RUNTIME` would do the trick for us

**Comment by karsaig** [comment]
:+1: to this
Would help with Pitest (aka PIT) also. Not to mutate generated code.

**Comment by elucash** [comment]
Introducing runtime retained (should cover class retention case also) `org.immutables.value.Generated`. It's a top-level annotation to avoid messing with fully qualified form (i.e. `..Value.Generated` vs `..Value$Generated`. Will be available in next release 2.6 being prepared
